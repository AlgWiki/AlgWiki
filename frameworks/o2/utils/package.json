{
  "name": "@oxy2/utils",
  "version": "0.0.1",
  "description": "Internal utilities for the O2 framework.",
  "keywords": [],
  "homepage": "https://github.com/AlgWiki/AlgWiki#readme",
  "repository": {
    "type": "git",
    "url": "https://github.com/AlgWiki/AlgWiki.git",
    "directory": "frameworks/o2/utils"
  },
  "bugs": {
    "url": "https://github.com/AlgWiki/AlgWiki/issues"
  },
  "author": "jakzo",
  "license": "MIT",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org"
  },
  "engines": {
    "node": ">=14"
  },
  "devDependencies": {
    "@types/node": "14.14.37"
  },
  "dependencies": {
    "tslib": "^2.2.0"
  }
}
