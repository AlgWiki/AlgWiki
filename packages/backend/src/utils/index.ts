export * from './context';
export * from './logger';
