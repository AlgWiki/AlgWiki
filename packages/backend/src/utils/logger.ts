import pino from 'pino';

export const baseLogger = pino({ base: null });
