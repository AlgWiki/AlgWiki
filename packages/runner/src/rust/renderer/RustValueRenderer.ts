import dedent from "dedent";

import { InferredPrimitive, KeyPair, Primitive, Renderer } from "../../Type";
import { RustTypeRenderer } from "./RustTypeRenderer";

export class RustValueRenderer implements Renderer {
  private readonly typeTemplater: RustTypeRenderer = new RustTypeRenderer();

  single<P extends Primitive>(t: P, v: InferredPrimitive<P>): string {
    switch (t) {
      case Primitive.Integer:
      case Primitive.Boolean:
        return JSON.stringify(v);
      case Primitive.Float: {
        let s = JSON.stringify(v);
        // Rust requires the decimal separator for float literals
        if (!s.includes(".")) {
          s += ".";
        }
        return s;
      }
      case Primitive.String:
        return `r###"${v}"###`;
      default:
        throw new Error(`Encountered unknown primitive of value: ${t}`);
    }
  }

  list<P extends Primitive>(t: P, v: InferredPrimitive<P>[]): string {
    return `vec![${v.map((u) => this.single(t, u))}]`;
  }

  linkedList<P extends Primitive>(t: P, v: InferredPrimitive<P>[]): string {
    const tType = this.typeTemplater.single(t);
    return dedent(`{
      use std::collections::LinkedList;
      let mut inner: LinkedList<${tType}> = LinkedList::new();
      ${v.map((u) => `inner.push_back(${this.single(t, u)});`).join("\n")}
      inner
    }`);
  }

  dictionary<K extends Primitive, V extends Primitive>(
    t: KeyPair<K, V>,
    v: Map<InferredPrimitive<K>, InferredPrimitive<V>>
  ): string {
    const kType = this.typeTemplater.single(t.key);
    const vType = this.typeTemplater.single(t.value);
    return dedent(`{
      use std::collections::HashMap;
      let mut inner: HashMap<${kType}, ${vType}> = HashMap::new();
      ${[...v.entries()]
        .map(
          ([kVal, vVal]) =>
            `inner.insert(${this.single(t.key, kVal)}, ${this.single(
              t.value,
              vVal
            )});`
        )
        .join("\n")}
      inner
    }`);
  }
}
