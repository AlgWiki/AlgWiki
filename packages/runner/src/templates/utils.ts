import { Language } from "@alg-wiki/types";
import { snakeCase } from "change-case";

import { RustTypeRenderer } from "./rust";
import { TypeRenderer } from "./types";

/**
 * A map of `TypeRenderer`s for each language (if any)
 */
export const LanguageTypeRenderer: Record<Language, TypeRenderer | null> = {
  [Language.Python3]: null,
  [Language.Rust]: new RustTypeRenderer(),
};

/**
 * Functions to create a valid identifier for each language
 * TODO: need to ensure these functions exclude language keywords, etc
 */
export const LanguageIdentifier: Record<Language, (inp: string) => string> = {
  [Language.Rust]: (inp) =>
    `${/[^a-z]/.test(inp[0]) ? "_" : ""}${snakeCase(inp)}`,
  [Language.Python3]: (inp) =>
    `${/[^a-z]/.test(inp[0]) ? "_" : ""}${snakeCase(inp)}`,
};

/**
 * The filename for the file containing the runner, which will be created in the
 * bind mount directory
 */
export const LanguageFileName: Record<Language, string> = {
  [Language.Python3]: "index.py",
  [Language.Rust]: "main.rs",
};
