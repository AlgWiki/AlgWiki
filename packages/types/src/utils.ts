import { snakeCase } from "change-case";

import { EncodedResultError, Language } from "./types";

export function isObject(x: unknown): x is Record<string, unknown> {
  // If `x` is already an object, it's returned. Otherwise it's wrapped.
  // Therefore this will check if a value is an object, and also rule out null/undefined.
  return x === new Object(x);
}

// TODO: doesn't perform a deep check
export function isEncodedResultError(
  x: unknown,
  boundaryKey: string
): x is EncodedResultError {
  return isObject(x) && boundaryKey in x;
}

export const LanguageCasing: Record<Language, (name: string) => string> = {
  [Language.Python3]: snakeCase,
};
