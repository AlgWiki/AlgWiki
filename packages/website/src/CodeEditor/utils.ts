export const isCodeMirrorRequired = (language: string): boolean => {
  const isMobile = false;
  const isLanguageSupportedByMonaco = language === "javascript";
  return isMobile || !isLanguageSupportedByMonaco;
};
