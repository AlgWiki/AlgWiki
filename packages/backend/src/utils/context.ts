import { createNamespace } from 'cls-hooked';
import { Logger } from 'pino';
import { baseLogger } from './logger';

export interface RequestContext {
  log: Logger;
}

export const ctxNs = createNamespace('request');

export const runWithCtx = <T>(ctx: RequestContext, cb: () => T) =>
  ctxNs.runAndReturn(() => {
    ctxNs.active.ctx = ctx;
    return cb();
  });

export const getCtx = () => ctxNs.active.ctx as RequestContext | undefined;

export const defaultCtx: RequestContext = {
  log: baseLogger,
};

/** Request context object with defaults when context has not been initialized. */
export const ctx = new Proxy(defaultCtx, {
  get(target, key) {
    const reqCtx = getCtx() || target;
    return (reqCtx as any)[key];
  },
});
