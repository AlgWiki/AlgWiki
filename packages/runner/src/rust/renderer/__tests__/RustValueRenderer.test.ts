import { RustValueRenderer } from "../RustValueRenderer";
import { typeMatrix } from "./utils";

describe("RustValueRenderer", () => {
  it("can render all values", () => {
    const renderedTypes = typeMatrix
      .map((type) => type.render(new RustValueRenderer()))
      .join("\n");
    expect(renderedTypes).toMatchSnapshot();
  });
});
