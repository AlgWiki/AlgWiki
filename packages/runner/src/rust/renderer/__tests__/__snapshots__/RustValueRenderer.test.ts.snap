// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RustValueRenderer can render all values 1`] = `
"42
vec![42]
{
  use std::collections::LinkedList;
  let mut inner: LinkedList<i64> = LinkedList::new();
  inner.push_back(42);
  inner
}
{
  use std::collections::HashMap;
  let mut inner: HashMap<i64, f64> = HashMap::new();
  inner.insert(42, 1.337);
  inner
}
{
  use std::collections::HashMap;
  let mut inner: HashMap<i64, &str> = HashMap::new();
  inner.insert(42, r###\\"1986\\"###);
  inner
}
{
  use std::collections::HashMap;
  let mut inner: HashMap<i64, bool> = HashMap::new();
  inner.insert(42, true);
  inner
}
1.337
vec![1.337]
{
  use std::collections::LinkedList;
  let mut inner: LinkedList<f64> = LinkedList::new();
  inner.push_back(1.337);
  inner
}
{
  use std::collections::HashMap;
  let mut inner: HashMap<f64, i64> = HashMap::new();
  inner.insert(1.337, 42);
  inner
}
{
  use std::collections::HashMap;
  let mut inner: HashMap<f64, &str> = HashMap::new();
  inner.insert(1.337, r###\\"1986\\"###);
  inner
}
{
  use std::collections::HashMap;
  let mut inner: HashMap<f64, bool> = HashMap::new();
  inner.insert(1.337, true);
  inner
}
r###\\"1986\\"###
vec![r###\\"1986\\"###]
{
  use std::collections::LinkedList;
  let mut inner: LinkedList<&str> = LinkedList::new();
  inner.push_back(r###\\"1986\\"###);
  inner
}
{
  use std::collections::HashMap;
  let mut inner: HashMap<&str, i64> = HashMap::new();
  inner.insert(r###\\"1986\\"###, 42);
  inner
}
{
  use std::collections::HashMap;
  let mut inner: HashMap<&str, f64> = HashMap::new();
  inner.insert(r###\\"1986\\"###, 1.337);
  inner
}
{
  use std::collections::HashMap;
  let mut inner: HashMap<&str, bool> = HashMap::new();
  inner.insert(r###\\"1986\\"###, true);
  inner
}
true
vec![true]
{
  use std::collections::LinkedList;
  let mut inner: LinkedList<bool> = LinkedList::new();
  inner.push_back(true);
  inner
}
{
  use std::collections::HashMap;
  let mut inner: HashMap<bool, i64> = HashMap::new();
  inner.insert(true, 42);
  inner
}
{
  use std::collections::HashMap;
  let mut inner: HashMap<bool, f64> = HashMap::new();
  inner.insert(true, 1.337);
  inner
}
{
  use std::collections::HashMap;
  let mut inner: HashMap<bool, &str> = HashMap::new();
  inner.insert(true, r###\\"1986\\"###);
  inner
}"
`;
